lang/py/src/avro/io.py

```diff
@@ -112,58 +131,123 @@ def _is_timezone_aware_datetime(dt):
+            type(datum)
+          )
+        )
+    return ValidationResult(
+      None if (isinstance(datum, (int, long)) and LONG_MIN_VALUE <= datum <= LONG_MAX_VALUE) else \
+      BASE_ERR.format('long', type(datum))
+    )
   elif schema_type in ['float', 'double']:
-    return isinstance(datum, (int, long, float))
+    return ValidationResult(
+      None if isinstance(datum, (int, long, float)) else \
+      BASE_ERR.format('(float, double)', type(datum))
+    )
   # Check for int, float, long and decimal
   elif schema_type == 'fixed':
     if (hasattr(expected_schema, 'logical_type') and
         expected_schema.logical_type == constants.DECIMAL):
-      return isinstance(datum, Decimal)
-    return isinstance(datum, str) and len(datum) == expected_schema.size
+      return ValidationResult(
+        None if isinstance(datum, Decimal) else \
+        BASE_ERR.format('logicalType(Decimal), baseType(fixed)', type(datum))
+      )
+    if not isinstance(datum, str):
+      return ValidationResult(BASE_ERR.format('str', type(datum)))
+    elif len(datum) != expected_schema.size:
+      return ValidationResult(
+        'Fixed type with invalid size, expected "{}" bytes found "{}" bytes'.format(
+          expected_schema.size, len(datum)
+        )
+      )
+    else:
+      return ValidationResult()
   elif schema_type == 'enum':
-    return datum in expected_schema.symbols
+    return ValidationResult(
+      None if datum in expected_schema.symbols else \
+      '"{}" is not among expected symbols {}'.format(datum, expected_schema.symbols)
+    )
   elif schema_type == 'array':
-    return (isinstance(datum, list) and
-      False not in [validate(expected_schema.items, d) for d in datum])
+    if not isinstance(datum, list):
+      return ValidationResult(BASE_ERR.format('list', type(datum)))
+    arr_items = [validate(expected_schema.items, d) for d in datum]
+    if not any(arr_items):
+      return ValidationResult()
+    return ValidationResult('\n'.join([e.msg for e in filter(lambda a:a, arr_items)]))
   elif schema_type == 'map':
-    return (isinstance(datum, dict) and
-      False not in [isinstance(k, basestring) for k in datum.keys()] and
-      False not in
-        [validate(expected_schema.values, v) for v in datum.values()])
+    if not isinstance(datum, dict):
+      return ValidationResult(BASE_ERR.format('dict', type(datum)))
+    if False in [isinstance(k, basestring) for k in datum.keys()]:
+      return ValidationResult('Map keys must be string typed')
+    val_items = [validate(expected_schema.values, v) for v in datum.values()]
+    if not any(val_items):
+      return ValidationResult()
+    return ValidationResult('\n'.join([e.msg for e in filter(lambda a:a, val_items)]))
   elif schema_type in ['union', 'error_union']:
-    return True in [validate(s, datum) for s in expected_schema.schemas]
+    if not all([validate(s, datum) for s in expected_schema.schemas]):
```

Comment:
nitpick: it's probably easier to read if using `if any(...)` instead of `not all(...)` :)